tests/integration/helpers.py

```diff
@@ -257,3 +257,81 @@ async def secondary_mongo_uris_with_sync_delay(ops_test: OpsTest, rs_status_data
 def generate_collection_id() -> str:
     new_id = "".join(choices(ascii_lowercase + digits, k=4)).replace("_", "")
     return f"collection_{new_id}"
+
+
+async def get_application_relation_data(
+    ops_test: OpsTest,
+    application_name: str,
+    relation_name: str,
+    key: str,
+    relation_id: str = None,
+    relation_alias: str = None,
+) -> Optional[str]:
+    """Get relation data for an application.
+
+    Args:
+        ops_test: The ops test framework instance
+        application_name: The name of the application
+        relation_name: name of the relation to get connection data from
+        key: key of data to be retrieved
+        relation_id: id of the relation to get connection data from
+        relation_alias: alias of the relation (like a connection name)
+            to get connection data from
+    Returns:
+        the that that was requested or None
+            if no data in the relation
+    Raises:
+        ValueError if it's not possible to get application unit data
+            or if there is no data for the particular relation endpoint
+            and/or alias.
+    """
+    unit_name = f"{application_name}/0"
+    raw_data = (await ops_test.juju("show-unit", unit_name))[1]
+
+    if not raw_data:
+        raise ValueError(f"no unit info could be grabbed for {unit_name}")
+    data = yaml.safe_load(raw_data)
+
+    # Filter the data based on the relation name.
+    relation_data = [v for v in data[unit_name]["relation-info"] if v["endpoint"] == relation_name]
+
+    if relation_id:
+        # Filter the data based on the relation id.
+        relation_data = [v for v in relation_data if v["relation-id"] == relation_id]
+
+    if relation_alias:
+        # Filter the data based on the cluster/relation alias.
+        relation_data = [
+            v
+            for v in relation_data
+            if json.loads(v["application-data"]["data"])["alias"] == relation_alias
+        ]
+
+    if len(relation_data) == 0:
+        raise ValueError(
+            f"no relation data could be grabbed on relation with endpoint {relation_name} and alias {relation_alias}"
+        )
+
+    return relation_data[0]["application-data"].get(key)
+
+
+async def get_secret_id(ops_test, app_or_unit: Optional[str] = None) -> str:
+    """Retrieve secert ID for an app or unit."""
+    complete_command = "list-secrets"
+    # if app_or_unit:
```

Comment:
Is there a reason why it is commented?

Reply:
Yeah... It doesn't work ;-)
I'm already discussing it with the Juju team (and manually filter in this code instead).

I'm finalizing this part after it's clear whether the Juju command could do the job, or manual filter remains.

Reply:
so should we wait for the juju team response? I would prefer not to have commented code in the tests :)

Reply:
Juju team have already responded, I just didn't yet have the time to check their response.

Leaving comments in the code has never been the plan. But either to remove them, or to use them replacing the at-the-time working solution.

[Suggestion](https://chat.charmhub.io/charmhub/pl/6wqy9hbfcbb5ufs6mm9bkfrz3e) from the Juju team worked, the code is adjusted accordingly.